### PR TITLE
codebreaker-logics

### PR DESCRIPTION
**Pull request by Nikdill**

### CODE REVIEW

File: codebreaker.gemspec [hunk 1]
```diff
@@ -1,27 +1,28 @@
...
+# frozen_string_literal: true
+
+lib = File.expand_path('lib', __dir__)
 $LOAD_PATH.unshift(lib) unless $LOAD_PATH.include?(lib)
-require "codebreaker/version"
+require 'codebreaker/version'
 
 Gem::Specification.new do |spec|
-  spec.name          = "codebreaker"
+  spec.name          = 'codebreaker'
```
Comment (vatnajokull) [inline codebreaker.gemspec]: You need uniq name for publishing this gem to RubyGems

File: lib/codebreaker/code_formatter.rb [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  private
+
+  def empty?(maker_code, user_code)
+    maker_code.empty? && user_code.empty?
+  end
+
+  def minus?(first_code, second_code)
+    index = first_code.index(second_code[0])
+    index && first_code[index] != second_code[index]
+  end
```
Comment (vatnajokull) [inline lib/codebreaker/code_formatter.rb]: Can we replace these methods with one method?
```
def format(maker_code, user_code)
  user_code.map.with_index do |char, index|
    if __some_check__
       _do_something__
       '+'
    elsif __some_check__
       _do_something__
       '-'
     else
       next
     end
  end
end
```

File: lib/codebreaker/codebreaker.rb [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+                    difficulty: :easy },
+            middle: { attempts: 10,
+                      hints: 1,
+                      difficulty: :middle },
+            hell: { attempts: 5,
+                    hints: 1,
+                    difficulty: :hell } }.freeze
+
+  def self.stats_list
+    CodeBreaker.stats || []
```
Comment (vatnajokull) [inline lib/codebreaker/codebreaker.rb]: Can we replace it with `self.stats`?

File: lib/codebreaker/codebreaker.rb [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+  def check_code(code)
+    @codemaker.verify(code)
+    dicrement_attemts unless win?
+    save_stats if game_end?
+    @codemaker.verified_code
+  end
+
+  def hint
+    return @codemaker.code[code_index] if hint?
```
Comment (vatnajokull) [inline lib/codebreaker/codebreaker.rb]: You don't need to add `return` in this case
If you have several lines, `return` stops execution and returns some value

File: lib/codebreaker/codebreaker.rb [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+  end
+
+  def code_index
+    index = rand(0..3)
+    return code_index if @hint_cache.include?(index)
+
+    @hint_cache << index
+    @hints -= 1
+    index
```
Comment (vatnajokull) [inline lib/codebreaker/codebreaker.rb]: Is it possible to use some algorithm without recursion?
Maybe we can delete used hints from secret code and shuffle result for making the result more random?

File: lib/codebreaker/codebreaker.rb [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+  end
+
+  private
+
+  def dicrement_attemts
+    @attempts -= 1
+  end
+
+  def save_stats
+    all_stats = CodeBreaker.stats || []
```
Comment (vatnajokull) [inline lib/codebreaker/codebreaker.rb]: can we replace with `all_stats = self.stats_list`?

File: lib/codebreaker/codemaker.rb [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+  end
+
+  private
+
+  def generate_random_code
+    4.times.map { rand(1..6) }
+  end
+
+  def format(user_code)
+    @verified_code = CodeFormatter.call(code.dup, code_to_array(user_code))
```
Comment (vatnajokull) [inline lib/codebreaker/codemaker.rb]: You can pass just `code` to your service

File: spec/code_breaker_spec.rb [hunk 8]
```diff
@@ -0,0 +1,111 @@
+PATH_TEST_STATS = './spec/stats_test.yml'.freeze
+
+RSpec.describe CodeBreaker do
+  subject(:codebreaker) { described_class.new('Alisa', :easy) }
+
+  after { File.delete(PATH_TEST_STATS) if File.exist?(PATH_TEST_STATS) }
+
+  let(:hell) { described_class.new('Test', :hell) }
+
+  let(:expected_sats) do
```
Comment (vatnajokull) [inline spec/code_breaker_spec.rb]: `expected_stats`

File: lib/codebreaker/codemaker.rb (hunk 7 above)
Comment (Nikdill) [reply lib/codebreaker/codemaker.rb]: If I don’t copy the object, then the formatting process changes the "specific code"

File: lib/codebreaker/code_formatter.rb [hunk 9]
```diff
@@ -12,25 +12,26 @@ def call
...
+        maker_code[index] = nil
+        '+'
+      else
+        char_index = maker_code.index(char)
+        if char_index && minus?(char_index, maker_code, user_code)
+          maker_code[char_index] = nil
+          '-'
+        end
+      end
+    end.compact.sort
```
Comment (vatnajokull) [inline lib/codebreaker/code_formatter.rb]: Is it possible to replace this nested checks to one `elsif` and one `else`?
```
if some_check
  do_some_action
  '+'
elsif another_check
  do_another_action
  '-'
else 
  next
end
```

File: spec/code_maker_spec.rb [hunk 10]
```diff
@@ -76,4 +76,14 @@
...
   end
+
+  context 'when "verify" work as expected and the code does not change' do
+    let(:codemaker_code_1) { described_class.new }
+
+    before do
+      allow(codemaker_code_1).to receive(:code).and_return([1, 2, 3, 4])
+    end
+
+    specify { expect { codemaker_code_1.verify(1234) }.not_to change(codemaker_code_1, :code) }
```
Comment (vatnajokull) [inline spec/code_maker_spec.rb]: It's better to use the following syntax
`expect { something }.to change(object, :attribute).by(0)`

**Pull request merged**